how-does-the-development-mode-work.md

```diff
@@ -0,0 +1,474 @@
+  <summary>
+    有一件事情有可能依然在迷惑你, 在Github 上面的 React 源码, 你将看到 __DEV__ 还在作为魔术变量来使用. 但是在 npm 的 React 代码, 其使用 process.env.NODE_ENV. 这是怎么工作的?
+  </summary>
+  One thing that might still confuse you is that in React source code on GitHub, you’ll see __DEV__ being used as a magic variable. But in the React code on npm, it uses process.env.NODE_ENV. How does that work?
+</details>
+
+<details>
+  <summary>
+    在历史上, 我们在源码中使用 __DEV__ 来满足 Facebook 的源码. 很长时间, React 是直接 copy 进入 Facebook 的代码库, 所以它需要跟随相同的规则. 对于 npm, 我们有一个构建步骤, 其按字面的在发布之前将 __DEV__ 检查替换成 process.env.NODE_ENV !== 'production'.
+  </summary>
+  Historically, we’ve used __DEV__ in the source code to match the Facebook source code. For a long time, React was directly copied into the Facebook codebase, so it needed to follow the same rules. For npm, we had a build step that literally replaced the __DEV__ checks with process.env.NODE_ENV !== 'production' right before publishing.
+</details>
+
+<details>
+  <summary>
+    这有时候是一个问题. 有时, npm 上工作正常的一个代码模式依赖一些 Node.js 的约定, 但是破坏了 Facebook, 反之亦然.
+  </summary>
+  This was sometimes a problem. Sometimes, a code pattern relying on some Node.js convention worked well on npm, but broke Facebook, or vice versa.
+</details>
+
+<details>
+  <summary>
+    这意味着, 在 React 源码使用 if (__DEV__) 的时候, 我们实际上对每一个包生成两个 bundle. 一个已经用 __DEV__ = true 预编译, 另一个用 __DEV__ = false 预编译. 在 npm 中的每一个包的入口点决定导出哪一个.
+  </summary>
+  This means that while the React source code says if (__DEV__), we actually produce two bundles for every package. One is already precompiled with __DEV__ = true and another is precompiled with __DEV__ = false. The entry point for each package on npm “decides” which one to export.
+</details>
+
+<details>
+  <summary>
+    例如:
+  </summary>
+  For Example:
+</details>
+
+```js
+if (process.env.NODE_ENV === 'production') {
+  module.exports = require('./cjs/react.production.min.js');
+} else {
+  module.exports = require('./cjs/react.development.js');
+}
+```
+
+<details>
+  <summary>
+    并且这里是你的打包器将 <code>'development'</code> 或者是 <code>'production'</code> 作为字符串插入的唯一地方, 并且这里是你的压缩器摆脱 development-only <code>require</code> 的地方.
+  </summary>
+  And that’s the only place where your bundler will interpolate either <code>'development'</code> or <code>'production'</code> as a string, and where your minifier will get rid of the development-only <code>require</code>.
+</details>
+
+<details>
+  <summary>
+    react.production.min.js 和 react.development.js 两者都不在有任何的 process.env.NODE_ENV 检查. 这是很棒的, 因为当实际在 Node.js 中运行的时候, 访问 process.env 稍许缓慢. 提前在两个模式下编译 bundle 也让我们优化文件体积更加一致, 无论你使用的是那种打包器和压缩器.
+  </summary>
+  Both react.production.min.js and react.development.js don’t have any process.env.NODE_ENV checks anymore. This is great because when actually running on Node.js, accessing process.env is somewhat slow. Compiling bundles in both modes ahead of time also lets us optimize the file size much more consistently, regardless of which bundler or minifier you are using.
```

Comment:
```suggestion
  Both react.production.min.js and react.development.js don’t have any <code>process.env.NODE_ENV</code> checks anymore. This is great because when actually running on Node.js, accessing <code>process.env</code> is somewhat slow. Compiling bundles in both modes ahead of time also lets us optimize the file size much more consistently, regardless of which bundler or minifier you are using.
```